internal/wordvault/server.go

```diff
@@ -933,6 +935,106 @@ func (s *Server) EditFsrsParameters(ctx context.Context, req *connect.Request[pb
 	return connect.NewResponse(&pb.EditFsrsParametersResponse{}), nil
 }
 
+func (s *Server) AddDeck(ctx context.Context, req *connect.Request[pb.AddDeckRequest]) (
+	*connect.Response[pb.AddDeckResponse], error) {
+
+	user := auth.UserFromContext(ctx)
+	if user == nil {
+		return nil, unauthenticated("user not authenticated")
+	}
+
+	if req.Msg.Lexicon == "" {
+		return nil, invalidArgError("need a lexicon")
+	}
+
+	if req.Msg.Name == "" {
+		return nil, invalidArgError("need a name")
+	}
+
+	deck, err := s.Queries.AddDeck(ctx, models.AddDeckParams{
+		UserID:      int64(user.DBID),
+		LexiconName: req.Msg.Lexicon,
+		Name:        req.Msg.Name,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return connect.NewResponse(&pb.AddDeckResponse{
+		Deck: &pb.Deck{
+			Id:      deck.ID,
+			Name:    deck.Name,
+			Lexicon: deck.LexiconName,
+		}}), nil
+}
+
+func (s *Server) GetDecks(ctx context.Context, req *connect.Request[pb.GetDecksRequest]) (
+	*connect.Response[pb.GetDecksResponse], error) {
+
+	user := auth.UserFromContext(ctx)
+	if user == nil {
+		return nil, unauthenticated("user not authenticated")
+	}
+
+	decks, err := s.Queries.GetDecks(ctx, int64(user.DBID))
+
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &pb.GetDecksResponse{
+		Decks: make([]*pb.Deck, len(decks)),
+	}
+
+	for i := range decks {
+		resp.Decks[i] = &pb.Deck{
+			Id:      decks[i].ID,
+			Name:    decks[i].Name,
+			Lexicon: decks[i].LexiconName,
+		}
+	}
+
+	return connect.NewResponse(resp), nil
+}
+
+func (s *Server) EditDeck(ctx context.Context, req *connect.Request[pb.EditDeckRequest]) (
+	*connect.Response[pb.EditDeckResponse], error) {
+
+	user := auth.UserFromContext(ctx)
+	if user == nil {
+		return nil, unauthenticated("user not authenticated")
+	}
+
+	if req.Msg.Id == 0 {
+		return nil, invalidArgError("need a deck")
+	}
+
+	if req.Msg.Name == "" {
+		return nil, invalidArgError("need a name")
+	}
+
+	deck, err := s.Queries.EditDeck(ctx, models.EditDeckParams{
+		ID:   req.Msg.Id,
+		Name: req.Msg.Name,
```

Comment:
what do you think of preventing them from naming two decks the same thing?

Reply:
Oh good catch -- we should do that

Reply:
Done -- opted for an additional query rather than a unique constraint but happy to change

Reply:
@BenMusch I think a unique constraint might be a bit simpler and faster. Would it be possible to change it over?

Reply:
yep! Will do this eve

Reply:
Oh @domino14  -- just remembered the main motivation for not doing it was because of case-sentitivity. Thoughts on whether that merits using this approach? We can also add the DB-level unique constraint as a backup

Reply:
You should be able to set a unique constraint on lower(list_name) (and user_id) I believe. Would that work?

Reply:
From what I found (http://shuber.io/case-insensitive-unique-constraints-in-postgres/) it requires using an extension an changing the type itself to be case-insensitive in the DB

> Unfortunately, Postgres does now allow us to define a unique constraint using LOWER like:
>
> ALTER TABLE users ADD CONSTRAINT email_unique UNIQUE (LOWER(email));
> ERROR:  syntax error at or near "("  
> 
> But don't worry there's a workaround! Instead of using the text data type, we can use the citext (case insensitive text) type! First we need to enable the [citext](http://www.postgresql.org/docs/current/interactive/citext.html) extension

Reply:
Oh I see, it seems like it's supported as a unique index but not a constraint? Will play around a bit https://pgsql-general.postgresql.narkive.com/HswV7weL/how-to-create-a-case-insensitive-unique-constraint

Reply:
Oh interesting. In that case let's just keep it as is. I would just change the API so it trims any leading/trailing spaces from the list name.

Reply:
what's wrong with using unique index?
```sql
create unique index email_unique on users (lower(email));
```

Reply:
I think nothing -- I'll polish this up tonight!